### ISSUE TITLE
Real-time shim doesn't optimize well when using coils

### ISSUE THREAD

**Issue by 4rnaudB** [shimming-toolbox/shimming-toolbox#474]
### Checklist

- [x] I've given this issue a self-descriptive, meaningful title (e.g. *Installation failure: problem creating launchers*)
- [ ] I've described my environment
- [x] I've described the bug
- [x] I've described the steps to reproduce the behaviour
- [ ] I've included screenshots
- [ ] I've given this issue the correct labels

### Bug Description
Using a coil prevents the optimizer from finding a good solution. For example, using only the first order, coef_riro for the first slice would have 4 coefs of the order of 0.0001 while, if using coils, 15 (coil) + 4 (first order) coefs are 0. This phenomenon happens for many of the slices. 

### Reproduction Steps
To reproduce the problem, you can use branch `ab/rtShim` and debug the file b0shim.py. The files are too big to drop in this issue, but they can probably be reproduced with many field maps and coils.

**Comment by Aurelienpujolm** [comment]
After some research, I've found some interresting results. My first guess is that, in your case, we don't have to optimize at all with the mse criteria).
After calculating the stability factor (2.24e-5), I attempted to examine the residual mean squared error (MSE), mean absolute error (MAE), and standard deviation (STD). The obtained results are as follows:

Residuals MSE = 1.0 (For this calculation, I utilized the initial version of the code before any modifications were made this year, including the use of np.sum, in order to confirm that the observed value is not influenced by the changes introduced to the MSE function.)
residuals_mae = 186
residuals_std = 106
So because the residuals mse is already equal to 1, then no optimization is made 
See this comment 

            # scipy minimize expects the return value of the residual function to be ~10^0 to 10^1
            # --> aiming for 1 then optimizing will lower that. We are using an initial guess of 0s so that the
            # regularization on the currents has no affect on the output stability factor.

I will try to see then, why the residuals_mse is equal to 1, before the optimization

**Comment by po09i** [comment]
> aiming for 1 then optimizing will lower that.

The stability factor should set it to 1, then it will optimize and bring it lower

**Comment by Aurelienpujolm** [comment]
> > aiming for 1 then optimizing will lower that.
> 
> The stability factor should set it to 1, then it will optimize and bring it lower

Yes, you are right.  Other than that, I saw that the same thing is happening for all type of lsq optimizers (mse, mae, std). So it's not just a mse problem ( Can anyone else please confirm that?)

**Comment by po09i** [comment]
> So it's not just a mse problem

Why is this a problem? It's just an initial value which we are aiming to minimize

**Comment by Aurelienpujolm** [comment]
> > So it's not just a mse problem
> 
> Why is this a problem? It's just an initial value which we are aiming to minimize

I wanted to check first if it was only because of the mse optimization, but it's not. (other than that what you are saying is true).

**Comment by Aurelienpujolm** [comment]
My hypothesis is that the problem come from the significant difference between coil_mat and unshimmed (where unshimmed is much smaller).

In cases where the optimization does not work, unshimmed is around 0.0012, while coil_mat is about 26. Conversely, in cases where it does work, the results are closer to 0.008 for unshimmed and around 11 for coil_mat. This leads me to lean towards the idea that the larger the gap between the two, the higher the risk that the optimization won't work (if we want to minimize a + b@x, when b >> a, it's necessary to push x towards 0).

Hence, in my view, it might be worth considering adjusting the terms of coil_mat and unshimmed to achieve more consistent results.

Also, in cases where the terms are close to 0, I haven't been able to find a specific pattern that unites them all. The averages and arrangements of the matrices didn't show significant differences.
